api/debug.go

```diff
@@ -95,3 +130,241 @@ func (d *DebugAPI) TraceBlockByHash(
+	}
+
+	ledger, err := evm.NewRemoteLedger(exeClient.ExecutionDataRPCClient(), blockHeight)
+	if err != nil {
+		return nil, fmt.Errorf("could not create remote ledger for height: %d, with: %w", blockHeight, err)
+	}
+
+	return evm.NewBlockExecutor(
+		block,
+		ledger,
+		d.config.FlowNetworkID,
+		d.blocks,
+		d.receipts,
+		d.logger,
+	)
+}
+
+func applyStateOverrides(config *tracers.TraceCallConfig, stateDB types.StateDB) error {
+	if config == nil || config.StateOverrides == nil {
+		return nil
+	}
+
+	diff := *config.StateOverrides
+	for addr, account := range diff {
+		// Override account nonce.
+		if account.Nonce != nil {
+			stateDB.SetNonce(addr, uint64(*account.Nonce))
+		}
+		// Override account(contract) code.
+		if account.Code != nil {
+			stateDB.SetCode(addr, *account.Code)
+		}
+		// Override account balance.
+		if account.Balance != nil {
+			u256Balance, _ := uint256.FromBig((*big.Int)(*account.Balance))
+			currentBalance := stateDB.GetBalance(addr)
+			// If given balance is greater-or-equal than the current balance,
+			// we need to add the diff to the current balance. Otherwise,
+			// we need to sub the diff from the current balance.
+			if u256Balance.Cmp(currentBalance) >= 0 {
+				diff := u256Balance.Sub(u256Balance, currentBalance)
+				stateDB.AddBalance(addr, diff, tracing.BalanceChangeUnspecified)
+			} else {
+				diff := currentBalance.Sub(currentBalance, u256Balance)
+				stateDB.SubBalance(addr, diff, tracing.BalanceChangeUnspecified)
+			}
+		}
+		if account.State != nil && account.StateDiff != nil {
+			return fmt.Errorf("account %s has both 'state' and 'stateDiff'", addr.Hex())
+		}
+		// Replace entire state if caller requires.
+		if account.State != nil {
+			for key, value := range *account.State {
+				stateDB.SetState(addr, key, value)
+			}
+		}
+		// Apply state diff into specified accounts.
+		if account.StateDiff != nil {
+			for key, value := range *account.StateDiff {
+				stateDB.SetState(addr, key, value)
+			}
+		}
+	}
+
+	return stateDB.Commit(true)
+}
+
+func (d *DebugAPI) resolveBlockNumberOrHash(block *rpc.BlockNumberOrHash) (uint64, error) {
+	err := fmt.Errorf("%w: neither block number nor hash specified", errs.ErrInvalid)
+	if block == nil {
+		return 0, err
+	}
+	if number, ok := block.Number(); ok {
+		return d.resolveBlockNumber(number)
+	}
+
+	if hash, ok := block.Hash(); ok {
+		evmHeight, err := d.blocks.GetHeightByID(hash)
+		if err != nil {
+			return 0, err
+		}
+		return evmHeight, nil
+	}
+
+	return 0, err
+}
+
+func (d *DebugAPI) resolveBlockNumber(number rpc.BlockNumber) (uint64, error) {
+	height := number.Int64()
+
+	// if special values (latest) we return latest executed height
+	if height < 0 {
+		executed, err := d.blocks.LatestEVMHeight()
+		if err != nil {
+			return 0, err
+		}
+		height = int64(executed)
+	}
+
+	return uint64(height), nil
+}
+
+func tracerForReceipt(
+	config *tracers.TraceConfig,
+	receipt *models.Receipt,
```

Comment:
Currently the only usage of this always passes `receipt: nil`. Is there an intention to use this method in other contexts in the future?

Reply:
Yes, exactly. The `receipt` is needed when generating traces for transactions. I took this part from the other two PRs, that did a general re-work on the traces API, to use the remote register API.